src/lib.rs

```diff
@@ -101,6 +103,7 @@ struct Core {
 struct Slot<T> {
     value: UnsafeCell<MaybeUninit<T>>,
     state: AtomicUsize,
+    has_reader: AtomicBool,
```

Comment:
is it really necessary to add an entire additional `AtomicBool` (a whole word) to store what is essentially one bit of information? would it make sense to store this in the state by setting one bit in the `state` field instead? we could use the first bit to indicate if there is a reader, and store the generation of the slot in the remaining 63 bits, accessing the generation by shifting the value by 1.

this would have a couple advantages: one, it would decrease the size of each slot by a word --- the memory usage isn't a _huge_ deal, but it does scale with the size of the buffer, which could be meaningful. 

more importantly, though, the current approach stores two separate pieces of shared state in the `Slot` type, which are not both accessed atomically. this means there is a potential for racy behavior to occur when one value has been updated and the other has not yet been. if we store the presence of a reader as a single bit in the `state` field, both values are always read and written in a single atomic operation.

on the other hand, the approach i'm describing introduces some extra complexity to the code, since the presence of the reader field is not obvious on the struct definition and is instead hidden behind a bitfield...if there isn't a potential for racy behavior here, it may be better to keep this in a separate field.